### PR TITLE
AES Encrypt

### PR DESCRIPTION
**Pull request by isaacadams**
This is a work in progress. @Romulus10 need to review this with you to ensure I am on the right track.

### CODE REVIEW

**Comment by Romulus10** [comment]
I think it looks good. We'll need some tests to make sure it works.

File: src/aes_tools/mod.rs [hunk 1]
```diff
@@ -1,12 +1,98 @@
...
+    rand_bytes(&mut buf).unwrap();
+    buf
+}
+
+fn calculate_resize(size: usize) -> usize {
+    size + (16 - size % 16)
+}
+
+/// currently not working properly
+fn normalize_input<T: Copy>(data: Vec<T>) -> (usize, Vec<T>) {
```
[the remark this answers is not in the document]
Comment (isaacadams) [reply src/aes_tools/mod.rs:91]: I am doing some cleaning up now and have removed this.

File: src/aes_tools/mod.rs [hunk 2]
```diff
@@ -1,12 +1,73 @@
...
+    aes_crypt(key, iv, buffer, Mode::Encrypt)
+}
 
-pub fn generate_cipher() {}
+pub fn aes_decrypt(key: &AesKey, iv: Vec<u8>, ciphertext: Vec<u8>) -> Vec<u8> {
+    aes_crypt(key, iv, ciphertext, Mode::Decrypt)
+}
 
-pub fn encrypt() {}
+fn aes_crypt(key: &AesKey, mut iv: Vec<u8>, input: Vec<u8>, mode: Mode) -> Vec<u8> {
```
Comment (Romulus10) [inline src/aes_tools/mod.rs:63]: Nice code reuse.

File: src/aes_tools/tests.rs [hunk 3]
```diff
@@ -1 +1,46 @@
...
+    sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
+
+    let cipher: AESCipher = AESCipher::new();
+
+    let ciphertext: Vec<u8> = cipher.encrypt(&message);
+    let plaintext = cipher.decrypt(ciphertext);
+
+    let message_decoded = String::from_utf8(plaintext).expect("Found invalid UTF-8");
+
+    assert_eq!(message.trim_end(), message_decoded.trim_end());
```
Comment (Romulus10) [inline src/aes_tools/tests.rs:38]: `.trim_end()` - good find.

File: src/aes_tools/mod.rs [hunk 4]
```diff
@@ -1,12 +1,73 @@
...
 #[cfg(test)]
 mod tests;
 
-pub fn generate_key() {}
+mod padding;
+
+use padding::pad;
+
+struct AESCipher {
+    encrypt_key: AesKey,
```
[the remark this answers is not in the document]
Comment (isaacadams) [reply src/aes_tools/mod.rs:15]: yah.. I am confused by this as well -- when I was researching, I saw people using the library this way which is why I broke it out into an encrypt/decrypt key. I think it's worth looking into a bit more to figure out what is going on here.

File: src/aes_tools/tests.rs (hunk 3 above)
Comment (isaacadams) [reply src/aes_tools/tests.rs:38]: Do you think we should move that into the decrypt method itself so that it returns the original message?

Comment (Romulus10) [reply src/aes_tools/tests.rs:38]: Yeah that's a good idea

**Pull request merged**